[GLib] Crash in SleepDisablerGLib::acquireInhibitor() when g_get_prgname() returns NULL

kdwk opened the pull request:
#### 1876825e173cc445960445a11fa60d3286d70ab2
<pre>
[GLib] Crash in SleepDisablerGLib::acquireInhibitor() when g_get_prgname() returns NULL
<a href="https://bugs.webkit.org/show_bug.cgi?id=303623">https://bugs.webkit.org/show_bug.cgi?id=303623</a>

Reviewed by Michael Catanzaro and Adrian Perez de Castro.

According to the GLib API documentation, g_get_prgname() can return NULL (<a href="https://docs.gtk.org/glib/func.get_prgname.html).">https://docs.gtk.org/glib/func.get_prgname.html).</a> In SleepDisablerGLib.cpp:95, the result of this function is passed directly into g_variant_new(), resulting in assert string != NULL to fail and crash the WebProcess. Instead, it can fall back to the executable path. If that also fails, return early.
* Source/WebCore/PAL/pal/system/glib/SleepDisablerGLib.cpp:
(PAL::SleepDisablerGLib::acquireInhibitor):

Canonical link: <a href="https://commits.webkit.org/304162@main">https://commits.webkit.org/304162@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14eeb6ba1ede84d5b6fb8cd31549ab79ace3c02d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134583 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7041 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45819 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142108 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86537 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136453 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7643 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6897 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102865 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70147 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137530 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5352 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120624 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83665 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5209 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2825 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2707 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38764 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144801 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6716 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39346 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111265 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6790 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5634 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111547 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28333 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5047 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116898 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/60594 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6766 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35091 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6577 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70350 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6813 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6688 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebCore/PAL/pal/system/glib/SleepDisablerGLib.cpp:
```diff
@@ -91,8 +91,10 @@ void SleepDisablerGLib::acquireInhibitor()
+    } else {
+        const gchar* prgname = g_get_prgname();
+        parameters = g_variant_new("(ss)", prgname ? prgname : "", m_reason.utf8().data());
+    }
```
I think we could use `FileSystem::currentExecutablePath()` and extract the basename as a fallback, it tries check where the `/proc/self/exe` symlink points to (on Linux, at least):


```suggestion
    } else if (const gchar* prgname = g_get_prgname()) {
        parameters = g_variant_new("(ss)", prgname, m_reason.utf8().data());
    } else if (const auto executablePath = FileSystem::currentExecutablePath(); !executablePath.isNull()) {
        GUniquePtr<char> executableName(g_path_get_basename(executablePath.data()));
        parameters = g_variant_new("(ss)", executableName.get(), m_reason.utf8().data());
    } else {
        parameters = g_variant_new("(ss)", "", m_reason.utf8().data());
    } 
```

mcatanzaro commented on Source/WebCore/PAL/pal/system/glib/SleepDisablerGLib.cpp:
```diff
@@ -91,8 +92,13 @@ void SleepDisablerGLib::acquireInhibitor()
+        parameters = g_variant_new("(ss)", executableName.get(), m_reason.utf8().data());
     } else
-        parameters = g_variant_new("(ss)", g_get_prgname(), m_reason.utf8().data());
+        parameters = g_variant_new("(ss)", "", m_reason.utf8().data());
```
Looking at the [API docs](https://specifications.freedesktop.org/idle-inhibit/latest/#org.freedesktop.ScreenSaver.Inhibit) there's no clear evidence that an empty string should be allowed, so I wouldn't even attempt this. Even if it works in GNOME, it might not work in another desktop. Just return early here.

kdwk committed:
[GLib] Crash in SleepDisablerGLib::acquireInhibitor() when g_get_prgname() returns NULL
https://bugs.webkit.org/show_bug.cgi?id=303623

Reviewed by Michael Catanzaro and Adrian Perez de Castro.

According to the GLib API documentation, g_get_prgname() can return NULL (https://docs.gtk.org/glib/func.get_prgname.html). In SleepDisablerGLib.cpp:95, the result of this function is passed directly into g_variant_new(), resulting in assert string != NULL to fail and crash the WebProcess. Instead, it can fall back to the executable path. If that also fails, return early.
* Source/WebCore/PAL/pal/system/glib/SleepDisablerGLib.cpp:
(PAL::SleepDisablerGLib::acquireInhibitor):

Canonical link: https://commits.webkit.org/304162@main
